### DISCUSSION TITLE
[GSoC]Seeking Guidance for Project #83: Adding Fuzzing Tests to Manifold

### DISCUSSION THREAD

**Discussion by Divyansh200102** [elalish/manifold#756] [Q&A]
Dear community,
First and foremost, I would like to extend my apologies if this is not the appropriate forum for making such inquiries. If there's a more suitable place to discuss this topic, please feel free to direct me there.
I am eager to contribute to the Manifold project by taking on the task outlined in issue [#83](https://github.com/opencax/GSoC/issues/83), which involves adding fuzzing tests for both 2D and 3D operations. However, I am unsure about how to get started and how to proceed with preparing a proposal for Google Summer of Code (GSOC) for this project.
Could a mentor please guide me on how can I get started on making my GSOC Proposal for this project?@pca006132 I found your name in the ideas list that you might be potentially taking up this project as a mentor. Could you please guide me on how I can get started on making my proposal for GSOC  for this project?

**Comment by pca006132** [comment]
Hi, this is indeed the right place to ask (or in the GSoC issue you linked). I think I need to ask others that are more experienced with GSOC for comments about proposal. @elalish do you have any pointer for this?

**Reply by Divyansh200102** [reply]
> Hi, this is indeed the right place to ask (or in the GSoC issue you linked). I think I need to ask others that are more experienced with GSOC for comments about proposal. @elalish do you have any pointer for this?

Sure,thank you for the reply

**Reply by elalish** [reply]
@Divyansh200102, the idea is to propose to us how you plan to work on the problem. Do you know anything about fuzz testing? Or CAD, meshes, or manifolds? Not that it's necessary, just curious where you're coming from. Have you looked through any of our code base yet, particularly our tests? A good proposal is specific enough to prove you've done your homework and thought through how to split up the project into smaller pieces. And of course we welcome questions here, but the more specific you can make them, the better answers you'll get.

**Reply by Divyansh200102** [reply]
@elalish Fuzz testing is still new to me although I have gone through your codebase and I am starting to get a basic understanding of it .I have good understanding of  C++ and thought this project to be very intriguing so I thought to give it a try .With a bit of your help I am sure I will pick up fuzz testing on a good enough level to complete this project.

If i am thinking correctly particularly regarding this project with what I understand we basically have to:

- First we have to create a concise AST representing 2D and 3D CSG operations: union, intersection, difference, extrude, revolve.
- Employ AFL or libFuzzer for structured input generation. Use recursive domain exploration to test various CSG operation combinations and geometric shapes.
-  Introduce random perturbations to generated shapes to test for epsilon-validity, assessing algorithm robustness against imperfect input geometries.
- Develop functions for CSG operations, ensuring compatibility with 2D and 3D geometries.
- Run fuzz tests with defined AST and perturbed geometries. Monitor for crashes, failed assertions, and unexpected behavior.
- Analyze fuzz testing results iteratively, addressing identified issues to enhance algorithm robustness and correctness.
-  Document fuzz testing setup, including AST definition, input domain, and perturbation methods. Record findings and address any discovered bugs or vulnerabilities.

**Reply by pca006132** [reply]
Yeah, and in addition to this you probably need to write some function that test for invalid geometries. We do not require them to be fast, so it should not be too hard.

**Reply by Divyansh200102** [reply]
So, can you please tell me how can I get started regarding the GSOC proposal ? Should I directly start writing the proposal and share the drafts with your for advice? And if you don't mind I need to ask are there any other people who have approached to you regarding the proposal of this particular project? @pca006132

**Reply by pca006132** [reply]
We do not have very concrete requirement for the proposal, I think even the above points are sufficient (for me at least). Maybe you can have a look at https://google.github.io/gsocguides/student/writing-a-proposal

And yes, another person approached me about this project, but he/she is also interested in the `minGap` function project as well.

**Reply by Divyansh200102** [reply]
> We do not have very concrete requirement for the proposal, I think even the above points are sufficient (for me at least). Maybe you can have a look at https://google.github.io/gsocguides/student/writing-a-proposal
> 
> And yes, another person approached me about this project, but he/she is also interested in the `minGap` function project as well.

Okay,thank you for the information I will get back to you after preparing the first draft of proposal.

**Reply by Divyansh200102** [reply]
@pca006132 @elalish I have prepared the first draft for my GSOC proposal can you please take a look at it and suggest some improvements.[https://docs.google.com/document/d/1VWupeKn-omZ2KImu_XgQ7xuXkLzIhtYusSv9BZ3v1fk/edit?usp=sharing](https://docs.google.com/document/d/1VWupeKn-omZ2KImu_XgQ7xuXkLzIhtYusSv9BZ3v1fk/edit?usp=sharing)

**Reply by elalish** [reply]
That's a lot, thanks! One thing - we're not actually BRL-CAD, though they are kind enough to help us and several other projects organize for GSoC. Each of the projects is more or less independent, though we tend to take dependencies on each other.

**Reply by Divyansh200102** [reply]
> That's a lot, thanks! One thing - we're not actually BRL-CAD, though they are kind enough to help us and several other projects organize for GSoC. Each of the projects is more or less independent, though we tend to take dependencies on each other. 

Oh, sure I will change the name in the proposal please feel free to put some suggestions in the proposal if you feel something needs to be improved in the proposal according to you .It would be a great help for me and in turn increase the chance of proposal getting selected. Also I am going to be bit busy for about 2 weeks due to some university work so I won't have time to contribute to the manifold repository and since I have made no contributions to this particular repository in the past as well so in the community engagement section of the proposal  I have highlighted some of my major past contributions in other open source community so is that fine?Also do you recommend any changes in the Deliverables and scheduling section? @elalish
